engine/runtime/src/main/java/org/enso/interpreter/runtime/data/ArrayOverBuffer.java

```diff
@@ -0,0 +1,44 @@
+package org.enso.interpreter.runtime.data;
+
+import com.oracle.truffle.api.interop.InteropLibrary;
+import com.oracle.truffle.api.interop.TruffleObject;
+import com.oracle.truffle.api.library.ExportLibrary;
+import com.oracle.truffle.api.library.ExportMessage;
+import org.enso.interpreter.dsl.Builtin;
+import org.enso.interpreter.runtime.library.dispatch.MethodDispatchLibrary;
+
+import java.nio.ByteBuffer;
+
+@ExportLibrary(InteropLibrary.class)
+@Builtin(pkg = "mutable", stdlibName = "Standard.Base.Data.Array.Array_Over_Buffer")
+public class ArrayOverBuffer implements TruffleObject {
```

Comment:
I implemented this as @JaroslavTulach suggested, can you please take a look (together with @hubertp) to see if I didn't do something stupid here?

I'm a bit unsure of whether I should also add `MethodDispatchLibrary` here, as `Array` has it. It seemed not strictly necessary (all tests I want do pass currently) but I'm not sure if ignoring it does not introduce some hard to understand issue?

Also, shall I add Enso shadow definitions for `ArrayOverBuffer`? Ideally I think from Enso's perspective, both of these types should be seen as the same `Array` Enso type - the particular class being a particular implementation of that. But I'm not sure if this can be achieved easily? What do you think?

I guess this touches the whole discussion of revisiting arrays and having multiple possible underlying representations for them. Although it would be good if we would not block integrating this PR by getting the arrays into fully shape as this may be a much longer endeavour, so it would be cool if we can agree on a solution that can be integrated tomorrow and possibly decide to improve it later on if necessary.

Reply:
I tested the following script:
```
from Standard.Base import all

polyglot java import bar.Foo

foreign js make_js_array = """
    return [103, 101, 102, 100];

main =
    pure_enso = [40, 30]
    from_java = Vector.Vector Foo.makeJavaList
    from_js = Vector.Vector make_js_array
    bytebuffer = (enso_project.data / "test.txt").read_last_bytes 10

    [pure_enso, from_java, from_js, bytebuffer].each vec->
        IO.println vec
        IO.println vec.length
        IO.println (vec.map (+2))
        IO.println vec.sort
        IO.println ""
```

and all seems to work as expected.

I'm not sure if we shouldn't extend method dispatch for `ArrayOverBuffer`, but currently it should just behave as any foreign array we may also get - which seems to be good enough for now. In particular, it definitely solves the usecase we use this for and it is currently only an internal function anyway.